test/test_integration.py

```diff
@@ -1717,3 +1717,98 @@ def socket_handler(listener):
                   timeout=(10, 0.5))
 
         self.tear_down()
+
+    def test_adapter_close(self):
+        self.set_up(secure=False)
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+
+            # We should get the initial request.
+            data = b''
+            while not data.endswith(b'\r\n\r\n'):
+                data += sock.recv(65535)
+
+            # We need to send back a response.
+            resp = (
+                b'HTTP/1.1 201 No Content\r\n'
+                b'Server: socket-level-server\r\n'
+                b'Content-Length: 0\r\n'
+                b'Connection: close\r\n'
+                b'\r\n'
+            )
+            sock.send(resp)
+            sock.close()
+
+        self._start_server(socket_handler)
+
+        a = HTTP20Adapter()
+        s = requests.Session()
+        s.mount('http://', a)
+        r = s.get('http://%s:%s' % (self.host, self.port))
+        connections_before_close = list(a.connections.values())
+        s.close()
+
+        # check that connections cache is empty
+        assert not a.connections
+
+        # check that all connections are actually closed
+        assert (connections_before_close and
+                all(conn._sock is None for conn in connections_before_close))
```

Comment:
Let's split the first assertion out and call it when we get the connections initially.